### ISSUE TITLE
Put Error :  Code 4

### ISSUE THREAD

**Issue by krtooch** [theophilusx/ssh2-sftp-client#339]
Hello i have an issue with a code 4, on put function.

It run on mac OS, node 14.16, ssh2-SFTP-Client 7.0.1

i use the script to upload the string data as a xml file.

The script works fine with one specific server, but got error on other one.

`
export const upload = async (data: string, name: string, credential: SFTPCredentials, options?: { cipher?: string[] }) => {
	const sftp = new sftpClient();
	await sftp.connect({
		...credential,
		debug: console.log,
	})
	const readable = new Readable();
	readable.push(Buffer.from(data));
	readable.push(null);
	console.log('****************************WE ARE CONNECTED*******************************')
	await sftp.put(readable, name);
	sftp.end();
};
`

here is my error log

`
Outbound: Sending SERVICE_REQUEST (ssh-userauth)
Inbound: Received SERVICE_ACCEPT (ssh-userauth)
Outbound: Sending USERAUTH_REQUEST (none)
Inbound: Received USERAUTH_FAILURE (password,publickey,keyboard-interactive)
Client: none auth failed
Outbound: Sending USERAUTH_REQUEST (password)
Inbound: Received USERAUTH_SUCCESS
CLIENT[sftp]: getConnection: got connection - promise resolved
CLIENT[sftp]: getConnection: finally clause fired
CLIENT[sftp]: Removing listener from ready event
CLIENT[sftp]: getConnection: Removing temp event listeners
CLIENT[sftp]: getSftpChannel: Adding temp event listeners
CLIENT[sftp]: getSftpChannel: created promise
Outbound: Sending CHANNEL_OPEN (r:0, session)
Inbound: CHANNEL_OPEN_CONFIRMATION (r:0, s:0)
Outbound: Sending CHANNEL_REQUEST (r:0, subsystem: sftp)
Inbound: CHANNEL_WINDOW_ADJUST (r:0, 4194304)
Inbound: CHANNEL_SUCCESS (r:0)
Outbound: Sending CHANNEL_DATA (r:0, 9)
Inbound: CHANNEL_DATA (r:0, 120)
SFTP: Inbound: Received VERSION (v3, exts:newline@vandyke.com,vendor-id)
CLIENT[sftp]: getSftpChannel: SFTP channel established
CLIENT[sftp]: getSftpChannel: finally clause fired
CLIENT[sftp]: getSftpChannel: Removing temp event listeners
****************************WE ARE CONNECTED*******************************
CLIENT[sftp]: put: Adding temp event listeners
Outbound: Sending CHANNEL_DATA (r:0, 75)
SFTP: Outbound: Buffered OPEN
CLIENT[sftp]: put source is a stream
Inbound: CHANNEL_DATA (r:0, 76)
SFTP: Inbound: Received HANDLE (id:0)
Outbound: Sending CHANNEL_DATA (r:0, 84)
SFTP: Outbound: Buffered FSETSTAT
Inbound: CHANNEL_DATA (r:0, 44)
SFTP: Inbound: Received STATUS (id:1, 0, "The attributes were set")
Outbound: Sending CHANNEL_DATA (r:0, 23907)
SFTP: Outbound: Sent WRITE (id:2)
Inbound: CHANNEL_DATA (r:0, 78)
SFTP: Inbound: Received STATUS (id:2, 4, "/BRUCE/INBOX/INV_q9I2VXUyBA_I_ETT_2021-08-07-12:08:86.xml")
Outbound: Sending CHANNEL_DATA (r:0, 76)
SFTP: Outbound: Buffered CLOSE
Inbound: CHANNEL_DATA (r:0, 78)
SFTP: Inbound: Received STATUS (id:3, 4, "/BRUCE/INBOX/INV_q9I2VXUyBA_I_ETT_2021-08-07-12:08:86.xml")
CLIENT[sftp]: put: Removing temp event listeners

`

The error i receive 

`put: /BRUCE/INBOX/INV_q9I2VXUyBA_I_ETT_2021-08-07-12:08:86.xml INBOX/INV_q9I2VXUyBA_I_ETT_2021-08-07-12:08:86.xml

{
    "code": 4,
    "custom": true
}

`

thanks for help

**Comment by theophilusx** [comment]
The log and error message looks odd. Not sure what is going on there.
However, I have a few comments and suggestions.

- When including code in your issue, please use the github code markup
syntax. If you just cut'n'paste, the code gets reformatted, which makes
it very hard to follow.

- The debug log doesn't look complete. Would be good to have a complete
debug log. Probably good to just attach it as a file as it may be quite
long.

- I don't use typescript, so had to modify your code. However, once
modified, I was not able to reproduce your issue or the odd looking
error message when attempting to upload to 3 different servers. The
error message should have some error text, not just the destination
filename. Don't know why yours doesn't have that.

- I would strongly recommend separating out the connect/end calls from
the function which uploads the data. Making a connection is a very slow
operation because of all the handshaking, key exchange and cipher
negotiation. You really only want to make as few connections as
possible, not a new one for every upload.

- Note that you don't have to use a read stream with put. You could just
  pass in your buffer of data directly. The put() method supports
  buffers, streams and file paths as input. 

- When using async/await, I think it is a good idea to wrap your code in
try/catch blocks. This can also enable you to add information to help
with debugging.

I have added some additional debug statements to the put method in the
master branch of the github repository. maybe these will provide some
additional info that will help track down the problem. Please try using
that version and see if it helps. (you can just update your package.json
file and put "theophilusx/ssh2-sftp-client" as the version. You will
need to remove package-lock.json and node_modules and then do an npm i)

I have also created a very simple sample script in the example directory
of the repository called stream-put.js. You might find it useful as an
example or to assist in debugging. It uses a dotenv .env file to get
credential information and you will need to change the require statement
for ssh2-sftp-client if you want to run it outside the repository.

The script expects a least one argument, the remote file path. Adding a
second argument will enable debugging. e.g.

node ./stream-pu.js './path/to/remote/file'

or

node ./stream-put.js './path/to/remote/file' debug

I would also try using the sftp command line and see if you can upload
data to the exact same path as you are specifying in your script. That
might provide some more details as I suspect your issue is with the
remote sftp server rather than the code. You may want to add one or more
-v options to make sftp more verbose.

**Comment by theophilusx** [comment]
Note that I'm about to push out version 7.0.2, which is based on a new ssh2 v1.3.0. That version fixes an issue affecting some timeout issues with handshaking which was particularly prevalent on macOS systems. I have also made some tweaks to error reporting which might help provide a better error message. Please try the new version to see if that helps.
